swift/extractor/mangler/SwiftMangler.cpp

```diff
@@ -1,89 +1,386 @@
-  // stamp all declarations with an id-ref of the containing module
-  auto moduleLabel = dispatcher.fetchLabel(decl.getModuleContext());
-  ret << moduleLabel;
-  if (decl.getKind() == swift::DeclKind::TypeAlias) {
-    // In cases like this (when coming from PCM)
-    //  typealias CFXMLTree = CFTree
-    //  typealias CFXMLTreeRef = CFXMLTree
-    // mangleAnyDecl mangles both CFXMLTree and CFXMLTreeRef into 'So12CFXMLTreeRefa'
-    // which is not correct and causes inconsistencies. mangleEntity makes these two distinct
-    // prefix adds a couple of special symbols, we don't necessary need them
-    ret << mangler.mangleEntity(&valueDecl);
+  auto parent = getParent(decl);
+  return initMangled(decl) << fetch(parent) << getExtensionIndex(decl, parent);
+}
+
+unsigned SwiftMangler::getExtensionIndex(const swift::ExtensionDecl* decl,
+                                         const swift::Decl* parent) {
+  // to avoid iterating multiple times on the parent of multiple extensions, we preload extension
+  // indexes once for each encountered parent into the `preloadedExtensionIndexes` mapping.
+  // Because we mangle declarations only once in a given trap/dispatcher context, we can safely
+  // discard preloaded indexes on use
+  if (auto found = preloadedExtensionIndexes.extract(decl)) {
+    return found.mapped();
+  }
+  if (auto parentModule = llvm::dyn_cast<swift::ModuleDecl>(parent)) {
+    llvm::SmallVector<swift::Decl*> siblings;
+    parentModule->getTopLevelDecls(siblings);
+    indexExtensions(siblings);
+  } else if (auto iterableParent = llvm::dyn_cast<swift::IterableDeclContext>(parent)) {
+    indexExtensions(iterableParent->getAllMembers());
   } else {
-    // prefix adds a couple of special symbols, we don't necessary need them
-    ret << mangler.mangleAnyDecl(&valueDecl, /* prefix = */ false);
+    assert(false && "non-local context must be module or iterable decl context");
   }
+  auto found = preloadedExtensionIndexes.extract(decl);
+  assert(found && "extension not found within parent");
+  return found.mapped();
+}
+
+void SwiftMangler::indexExtensions(llvm::ArrayRef<swift::Decl*> siblings) {
+  auto index = 0u;
+  for (auto sibling : siblings) {
+    if (sibling->getKind() == swift::DeclKind::Extension) {
+      preloadedExtensionIndexes.emplace(sibling, index);
+    }
+    ++index;
+  }
+}
+
+SwiftMangledName SwiftMangler::visitGenericTypeDecl(const swift::GenericTypeDecl* decl) {
+  auto ret = visitValueDecl(decl);
+  if (!ret) {
+    return {};
+  }
+  if (auto genericParams = decl->getParsedGenericParams()) {
+    ret << '<' << genericParams->size() << '>';
+  }
+  return ret;
+}
+
+SwiftMangledName SwiftMangler::visitAbstractTypeParamDecl(
+    const swift::AbstractTypeParamDecl* decl) {
+  return visitValueDecl(decl, /* force */ true);
+}
+
+SwiftMangledName SwiftMangler::visitGenericTypeParamDecl(const swift::GenericTypeParamDecl* decl) {
+  auto ret = visitAbstractTypeParamDecl(decl);
```

Comment:
Nit: Here and throughout, when we don't have control flow, I'd personally prefer to get rid of the variable and just write `return visitAbstractTypeParamDecl(decl) << '_' << decl->getDepth() ...`.

(And indeed reading further ahead, I see most of the functions are already structured that way.)

Reply:
yep, that was the idea, but it was introduced along the way and it seems some places where not touched by the refactoring. Will do!